src/impl_owned_array.rs

```diff
@@ -401,16 +402,88 @@ impl<A, D> Array<A, D>
     ///            [0., 0., 0., 0.],
     ///            [1., 1., 1., 1.]]);
     /// ```
-    pub fn push(&mut self, axis: Axis, array: ArrayView<A, D::Smaller>)
-        -> Result<(), ShapeError>
+    pub fn push(&mut self, axis: Axis, array: ArrayView<A, D::Smaller>) -> Result<(), ShapeError>
     where
         A: Clone,
         D: RemoveAxis,
     {
         // same-dimensionality conversion
-        self.append(axis, array.insert_axis(axis).into_dimensionality::<D>().unwrap())
+        self.append(
+            axis,
+            array.insert_axis(axis).into_dimensionality::<D>().unwrap(),
+        )
     }
 
+    /// Return the offset between the pointer to the beginning of the heap
+    /// allocated by Array (`self.data.ptr`) and `self.ptr`.
+    unsafe fn offset_from_data(&self) -> isize {
+        if std::mem::size_of::<A>() != 0 {
+            self.as_ptr().offset_from(self.data.as_ptr())
+        } else {
+            0
+        }
+    }
+
+    /// Convert from any stride to the default stride
+    pub fn to_default_stride(&mut self) {
+        let view_mut =
+            unsafe { ArrayViewMut::new(self.ptr, self.dim.clone(), self.strides.clone()) };
+        let offset = unsafe { self.offset_from_data() };
+        unsafe { self.ptr.offset(offset) };
+        self.strides = self.dim.default_strides();
+        let mut index_ = match self.dim.first_index() {
+            Some(x) => x,
+            None => unreachable!(),
+        };
+        let mut swap_idx: Vec<(isize, isize)> = Vec::new();
+        loop {
+            let self_index = self
+                .dim
+                .stride_offset_checked(&self.strides, &index_)
+                .unwrap();
+            let view_mut_index = view_mut
+                .dim
+                .stride_offset_checked(&view_mut.strides, &index_)
+                .unwrap()
+                + offset;
+            swap_idx.push((
+                std::cmp::min(self_index, view_mut_index),
+                std::cmp::max(self_index, view_mut_index),
+            ));
+
+            index_ = match self.dim.next_for(index_) {
+                Some(x) => x,
+                None => {
+                    break;
+                }
+            };
+        }
+        let swap_idx = swap_idx
+            .into_iter()
+            .collect::<HashSet<_>>()
+            .into_iter()
+            .collect::<Vec<_>>();
+        for (x, y) in swap_idx.iter() {
+            unsafe { mem::swap(self.ptr.offset(*x).as_mut(), self.ptr.offset(*y).as_mut()) };
+        }
+    }
+
+    /// Shrinks the capacity of the array as much as possible.
+    ///
+    /// ```
+    /// use ndarray::array;
+    /// use ndarray::s;
+    ///
+    /// let a = array![[0, 1, 2], [3, 4, 5], [6, 7,8]];
+    /// let mut a = a.slice_move(s![.., 0..2]);
+    /// let b = a.clone();
+    /// a.shrink_to_fit();
+    /// assert_eq!(a, b);
+    /// ```
+    pub fn shrink_to_fit(&mut self) {
+        self.to_default_stride();
```

Comment:
to default stride is an expensive operation - so we can probably avoid it on an array that is already perfectly fit.

We do prefer to not lean too strongly on a default memory order if possible. Any contiguous array should not have its memory order thrown around. I would even prefer if an array keeps its memory order the same, if it can.